synapse/handlers/room_summary.py

```diff
@@ -0,0 +1,285 @@
+
+    async def get_room_summary(
+        self,
+        requester: Optional[str],
+        room_id: str,
+        remote_room_hosts: List[str],
+    ) -> JsonDict:
+        """
+        Implementation of the room summary C-S API MSC3244
+
+        Args:
+            requester:  user id of the user making this request,
+                can be None for unauthenticated requests
+
+            room_id: room id to start the summary at
+
+            remote_room_hosts: a list of homeservers to try fetching data through
+                if we don't know it ourselves
+
+        Returns:
+            summary dict to return
+        """
+        is_in_room = await self._store.is_host_joined(room_id, self._server_name)
+
+        if is_in_room:
+            room_summary = await self._summarize_local_room(requester, None, room_id)
+
+            if requester:
+                membership, _ = await self._store.get_local_current_membership_for_user_in_room(
+                    requester, room_id
+                )
+
+                room_summary["membership"] = membership or "leave"
+        else:
+            room_summary = await self._summarize_remote_room(room_id, remote_room_hosts)
+
+            # TODO validate that the requester has permission to see this room
+            # https://github.com/matrix-org/matrix-doc/pull/3266/files#diff-97aeb566f3ce4bd6ec3b98e71ecbca3d6e86c0407e6a82afbc57e86bf0316607R106-R108
+
+        return room_summary
+
+    async def _summarize_local_room(
+        self,
+        requester: Optional[str],
+        origin: Optional[str],
+        room_id: str,
+    ) -> JsonDict:
+        """
+        Generate a room entry and a list of event entries for a given room.
+
+        Args:
+            requester:
+                The user requesting the summary, if it is a local request. None
+                if this is a federation request.
+            origin:
+                The server requesting the summary, if it is a federation request.
+                None if this is a local request.
+            room_id: The room ID to summarize.
+
+        Returns:
+            summary dict to return
+        """
+        if not await self._is_room_accessible(room_id, requester, origin):
+            return None
+
+        return await self._build_room_entry(room_id)
+
+    async def _summarize_remote_room(
+        self,
+        room_id: str,
+        remote_room_hosts: List[str],
+    ) -> JsonDict:
+        """
+        Request room entries and a list of event entries for a given room by querying a remote server.
+
+        Args:
+            room_id: The room to summarize.
+            remote_room_hosts: List of homeservers to attempt to fetch the data from.
+
+        Returns:
+            summary dict to return
+        """
+        logger.info("Requesting summary for %s via %s", room_id, remote_room_hosts)
+
+        return None  # TODO federation API
+
+    # TODO extract into mixin/helper method
```

Comment:
Might make sense to go on `synapse.api.auth.Auth`? Or could leave it here and make it public and have the spaces summary use the `RoomSummaryHandler`?